src/db/begin_transaction.cr

```diff
@@ -0,0 +1,26 @@
+module DB
+  module BeginTransaction
+    abstract def begin_transaction : Transaction
+
+    def transaction
+      tx = begin_transaction
+      begin
+        yield tx
+      rescue e
+        # TODO check if `unless tx.closed?` should be added.
+        # If it is not added, when an exception occurs after the
+        # transaction is closed (explicit rollback/commit)
+        # the call to rollback will trigger an exception.
+        # Since the code will be in between `#transaction(&block)`
+        # seems reasonable to raise an error since it might trigger
+        # a design flaw. Transaction might been commited already.
+        # Maybe we should wrap e in another exception that clarifies
+        # this scenario.
+        tx.rollback
```

Comment:
+1 to adding `unless tx.closed?`

If we're at this point, there has already been an exception thrown from the user's code, which will bubble and be catched by the client code. If the tx is already closed, it means that he manually commited or rolled it back, so it makes no sense to me to try and do that again, triggering an additional exception. The design flaw is exposed by the original exception, there's no need for another one.

@asterite maybe you want to break the tie? :-P